sdks/go/pkg/beam/core/util/jsonx/jsonx_test.go

```diff
@@ -0,0 +1,82 @@
+// Licensed to the Apache Software Foundation (ASF) under one or more
+// contributor license agreements.  See the NOTICE file distributed with
+// this work for additional information regarding copyright ownership.
+// The ASF licenses this file to You under the Apache License, Version 2.0
+// (the "License"); you may not use this file except in compliance with
+// the License.  You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package jsonx
+
+import (
+	"math"
+	"testing"
+)
+
+type someStruct struct {
+	SomeNumber  int64
+	SomeString  string
+	SomeBoolean bool
+}
+
+func (s *someStruct) equals(other *someStruct) bool {
+	return (s.SomeNumber == other.SomeNumber) && (s.SomeString == other.SomeString) && (s.SomeBoolean == other.SomeBoolean)
+}
+
+func TestEncodeDecode(t *testing.T) {
+	var tests = []struct {
+		name        string
+		inputNum    int64
+		inputString string
+		inputBool   bool
+	}{
+		{
+			"simple inputs",
+			10,
+			"information",
+			true,
+		},
+		{
+			"default inputs",
+			0,
+			"",
+			false,
+		},
+		{
+			"maximum int",
+			math.MaxInt64,
+			"big int",
+			true,
+		},
+		{
+			"minimum int",
+			math.MinInt64,
+			"little int",
+			false,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			inputStruct := &someStruct{SomeNumber: test.inputNum, SomeString: test.inputString, SomeBoolean: test.inputBool}
+			enc, err := Marshal(inputStruct)
+			if err != nil {
+				t.Fatalf("Marshal(%v) failed, got %v", inputStruct, err)
+			}
```

Comment:
This package is in a bit of a weird position in that it has already been earmarked for eventual deprecation once the SDK defaults to schemas. I couldn't tell you how far along we are to consider that complete, but it may be worth looking into. In the meantime, the json package has a `Valid()` function that checks if a byte slice is a valid JSON encoding. That should be straightforward, albeit somewhat unsatisfying given that we're really just making sure our dips into the json package's functions are working as intended.

Glad you caught the branch name, this has increasingly become my convention if there's a good name

Reply:
This package existed because internally to Google, we needed to migrate to the Go 1.16+ version of the JSON package prior to it becoming available, and abstracting it to a single package allowed all common uses to migrate in one place, rather than in the few places. It also allowed registering of the protojson package with the standard JSON encoder, so that protocol buffers actually get JSON encoded.

There's one place we do need to migrate to schemas before we get rid of the package though and that is DoFn serialization. It was never a quick enough change that I could prioritize it.

Reply:
(We'd also need to make sure it works for the variety of DoFns in use internally to Google as well, so it's not necessarily simple work.)